sdk/sync/client_topic.go

```diff
@@ -0,0 +1,176 @@
+// the payload type on the Topic; otherwise Publish will error.
+//
+// This method returns synchronously, once the item has been published
+// successfully, returning the sequence number of the new item in the ordered
+// topic, or an error if one ocurred, starting with 1 (for the first item).
+//
+// If error is non-nil, the sequence number must be disregarded.
+func (c *Client) Publish(ctx context.Context, topic *Topic, payload interface{}) (seq int64, err error) {
+	rp := c.extractor(ctx)
+	if rp == nil {
+		return -1, ErrNoRunParameters
+	}
+
+	log := c.log.With("topic", topic.name)
+	log.Debugw("publishing item on topic", "payload", payload)
+
+	if !topic.validatePayload(payload) {
+		err = fmt.Errorf("invalid payload type; expected: [*]%s, was: %T", topic.typ, payload)
+		return -1, err
+	}
+
+	// Serialize the payload.
+	bytes, err := json.Marshal(payload)
+	if err != nil {
+		err = fmt.Errorf("failed while serializing payload: %w", err)
+		return -1, err
+	}
+
+	log.Debugw("serialized json payload", "json", string(bytes))
+
+	key := topic.Key(rp)
+
+	log.Debugw("resolved key for publish", "key", key)
+
+	// Perform a Redis transaction, adding the item to the stream and fetching
+	// the XLEN of the stream.
+	args := new(redis.XAddArgs)
+	args.ID = "*"
+	args.Stream = key
+	args.Values = map[string]interface{}{RedisPayloadKey: bytes}
+
+	pipe := c.rclient.TxPipeline()
+	_ = pipe.XAdd(args)
+	xlen := pipe.XLen(key)
+
+	_, err = pipe.ExecContext(ctx)
+	if err != nil {
+		log.Debugw("failed to publish item", "error", err)
+		return -1, err
+	}
+
+	seq = xlen.Val()
+	c.log.Debugw("successfully published item; sequence number obtained", "seq", seq)
+
+	return seq, err
+}
+
+// MustPublish calls Publish, panicking if it errors.
+//
+// Suitable for shorthanding in test plans.
+func (c *Client) MustPublish(ctx context.Context, topic *Topic, payload interface{}) (seq int64) {
+	seq, err := c.Publish(ctx, topic, payload)
+	if err != nil {
+		panic(err)
+	}
+	return seq
+}
+
+// Subscribe subscribes to a topic, consuming ordered, typed elements from
+// index 0, and sending them to channel ch.
+//
+// The supplied channel must be buffered, and its type must be a value or
+// pointer type matching the topic type. If these conditions are unmet, this
+// method will error immediately.
+//
+// The caller must consume from this channel promptly; failure to do so will
+// backpressure the Client's subscription event loop.
+func (c *Client) Subscribe(ctx context.Context, topic *Topic, ch interface{}) (*Subscription, error) {
+	rp := c.extractor(ctx)
+	if rp == nil {
+		return nil, ErrNoRunParameters
+	}
+
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	if err := c.ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	chv := reflect.ValueOf(ch)
+	if k := chv.Kind(); k != reflect.Chan {
+		return nil, fmt.Errorf("value is not a channel: %T", ch)
+	}
+
+	if chv.Cap() == 0 {
```

Comment:
Why do we require the channel to be buffered?

Reply:
To avoid blocking the delivery event loop if the consumer is not reading. Now that you say it, I think this might cause a regression for some test plans.

Reply:
What so bad about blocking the delivery event loop if the consumer is not reading? Generally if a consumer subscribes to the sync service, it should be reading the values coming from the subscription... I am trying to come up with a scenarios where we would want to `subscribe` but ignore the values, and I can't come up with one - we'd be wasting memory in the channel for no good reason.

Reply:
Yup, I guess there's no harm done if we drop this requirement.